pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+	r.queue()
+	return nil
+}
+
+// Queues a reconcile.
+func (r *targetStatusReconciler) queue() {
+	r.ch <- event.GenericEvent{
+		Object: &appsv1.DaemonSet{},
+	}
+}
+
+// Reconcile polls the collector pods, fetches and aggregates target status and upserts
+// into each PodMonitoring's Status field.
+func (r *targetStatusReconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// minDelay := time.NewTimer(10 * time.Second)
+
+	minDelay := r.tickerBuilder()
+	// defer minDelay.Stop()
+
+	onceLog := new(sync.Once)
+
+	// Wrapper function so we could exit early.
+	err := func() error {
+		now := time.Now()
+		if err := poll(ctx, r.logger, r.opts, r.manager); err != nil {
+			// Not having data is not an error.
+			if apierrors.IsNotFound(err) {
+				onceLog.Do(func() {
+					r.logger.Error(err, "Kubernetes config not found")
+				})
+				return nil
+			}
+			return err
+		}
+		// In case Kubernetes config is removed and re-added.
+		onceLog = new(sync.Once)
+
+		duration := time.Since(now)
+		targetStatusDuration.WithLabelValues().Set(float64(duration.Milliseconds()))
+		return nil
+	}()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if we beat the timer, otherwise wait.
+	<-minDelay.Channel()
+	// Re-trigger event
+	r.queue()
+
+	return reconcile.Result{}, nil
+}
+
+func poll(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) error {
+	targets, err := fetchTargets(ctx, logger, opts, manager)
+	if err != nil {
+		return err
+	}
+
+	return populateTargets(ctx, logger, manager, targets)
+}
+
+func fetchTargets(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) ([]*prometheusTargetsResultAndTime, error) {
+	podMap, err := manager.getPrometheusPods(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up pod job queue and jobs
+	podDiscoveryChan := make(chan prometheusPod, opts.TargetPollConcurrency)
+	wg := sync.WaitGroup{}
+	wg.Add(int(opts.TargetPollConcurrency))
+
+	// Must be unbounded or else we deadlock.
+	targetChan := make(chan *prometheusTargetsResultAndTime)
+	for i := uint16(0); i < opts.TargetPollConcurrency; i++ {
+		// Wrapper function to handle errors.
+		go func(thread uint16) {
+			fetcher, err := manager.getTargetFetcher(ctx, thread)
+			if err != nil {
+				logger.Error(err, "error getting target fetcher")
+				return
+			}
+			err = processTargetQueue(ctx, logger, &wg, fetcher, podDiscoveryChan, targetChan)
+			if err != nil {
+				logger.Error(err, "error processing target queue")
+			}
+		}(i)
+	}
+
+	// Unbuffered channels are blocking so make sure we end the goroutine processing them.
+	go func() {
+		for port, pods := range podMap {
+			for _, pod := range pods {
+				podDiscoveryChan <- prometheusPod{
+					port: port,
+					pod:  pod,
+				}
+			}
+		}
+
+		// Must close so jobs aren't waiting on the channel indefinitely.
+		close(podDiscoveryChan)
+
+		// Close target after we're sure all targets are queued.
```

Comment:
I don't think these two lines need to be in the sending goroutine. They should be blocking the main thread execution until were done processing the pods. Then we can proceed with building and returning our results synchronously.

Reply:
This doesn't work because writing to an unbuffered channel is blocking on reading. The pod channel is read by the worker thread who writes to the target channel so it's blocked here. As an example, consider 2 worker threads and 3 pods. The worker threads start by waiting for work. Somebody needs to give the worker threads some work. Today this is the separate goroutine. The worker threads get 2 pods and they spit out 2 targets and they're blocked here. The main thread is reading targets to unblock the workers. The workers are unblocked and they get the next pod and spit out the next target. The main thread is still reading targets and is now blocked until the channel is closed. That's why the sending thread also closes the target channel because it knows that it sent all the pods, so all targets should be sent.

There may be a more clever way to do this with only two threads, but it's tricky because there's two channels involved.